core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java

```diff
@@ -233,25 +358,185 @@ public String getWriteableName() {
+        builder.accuracy = in.readFloat();
+        builder.sort = SortBy.PROTOTYPE.readFrom(in);
+        builder.stringDistance = StringDistanceImpl.PROTOTYPE.readFrom(in);
+        builder.maxEdits = in.readVInt();
+        builder.maxInspections = in.readVInt();
+        builder.maxTermFreq = in.readFloat();
+        builder.prefixLength = in.readVInt();
+        builder.minWordLength = in.readVInt();
+        builder.minDocFreq = in.readFloat();
+        return builder;
     }
 
     @Override
     protected boolean doEquals(TermSuggestionBuilder other) {
-        // NORELEASE
-        return false;
+        return Objects.equals(suggestMode, other.suggestMode) &&
+               Objects.equals(accuracy, other.accuracy) &&
+               Objects.equals(sort, other.sort) &&
+               Objects.equals(stringDistance, other.stringDistance) &&
+               Objects.equals(maxEdits, other.maxEdits) &&
+               Objects.equals(maxInspections, other.maxInspections) &&
+               Objects.equals(maxTermFreq, other.maxTermFreq) &&
+               Objects.equals(prefixLength, other.prefixLength) &&
+               Objects.equals(minWordLength, other.minWordLength) &&
+               Objects.equals(minDocFreq, other.minDocFreq);
     }
 
     @Override
     protected int doHashCode() {
-        // NORELEASE
-        return 0;
+        return Objects.hash(suggestMode, accuracy, sort, stringDistance, maxEdits, maxInspections,
+                            maxTermFreq, prefixLength, minWordLength, minDocFreq);
+    }
+
+
+    /** An enum representing the valid suggest modes. */
+    public enum SuggestMode implements Writeable<SuggestMode> {
+        /** Only suggest terms in the suggest text that aren't in the index. This is the default. */
+        MISSING,
+        /** Only suggest terms that occur in more docs then the original suggest text term. */
+        POPULAR,
+        /** Suggest any matching suggest terms based on tokens in the suggest text. */
+        ALWAYS;
+
+        protected static SuggestMode PROTOTYPE = MISSING;
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            out.writeVInt(ordinal());
+        }
+
+        @Override
+        public SuggestMode readFrom(final StreamInput in) throws IOException {
+            int ordinal = in.readVInt();
+            if (ordinal < 0 || ordinal >= values().length) {
+                throw new IOException("Unknown SuggestMode ordinal [" + ordinal + "]");
+            }
+            return values()[ordinal];
+        }
+
+        public static SuggestMode fromString(final String str) {
+            Objects.requireNonNull(str, "Input string is null");
+            return valueOf(str.toUpperCase(Locale.ROOT));
+        }
+
+        // TODO: move this to SuggestUtils when the builder becomes the intermediate representation
+        static SuggestMode fromUnderlying(final org.apache.lucene.search.spell.SuggestMode mode) {
```

Comment:
Also, we will need the oposite conversion from the ES enums (e.g. TermSuggestionBuilder.SuggestMode) to the Lucene enums (org.apache.lucene.search.spell.SuggestMode) used in the DirectSpellcheckerSettings later anyway, so rather than having `fromUnderlying` better turn the direction around to each enum knows how to produce the corresponding low-level enum.